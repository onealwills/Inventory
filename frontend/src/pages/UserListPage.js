import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { deleteUser, listUsers } from "../actions/userActions";
import LoadingBox from "../components/LoadingBox";
import MessageBox from "../components/MessageBox";
import { USER_DETAILS_RESET } from "../constants/userContants";

export default function UserListPage() {
  const userList = useSelector((state) => state.userList);
  const { loading, error, users } = userList;
  console.log("users list >>>", users);

  const userDelete = useSelector((state) => state.userDelete);
  const {
    loading: loadingDelete,
    error: errorDelete,
    success: successDelete,
  } = userDelete;
  console.log("loading delete >>>", userDelete);

  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(listUsers());
    dispatch({
      type: USER_DETAILS_RESET,
    });
  }, [dispatch, successDelete]);

  const deleteHandler = (user) => {
    if (window.confirm("you really wann delete that asshole")) {
      dispatch(deleteUser(user._id));
    }
  };
  return (
    <div>
      <h1>Users List</h1>
      {loadingDelete && <LoadingBox></LoadingBox>}
      {errorDelete && <MessageBox type="danger">{errorDelete}</MessageBox>}
      {successDelete && (
        <MessageBox type="success">User Deleted Successfully</MessageBox>
      )}
      {loading ? (
        <LoadingBox></LoadingBox>
      ) : error ? (
        <MessageBox type="danger">{error}</MessageBox>
      ) : (
        <div>
          <thead>
            <tr>
              <th>ID</th>
              <th>NAME</th>
              <th>EMAIL</th>
              <th>IS SUPER ADMIN</th>
              <th>IS ADMIN</th>
              <th>IS STOCK KEEPER</th>
            </tr>
          </thead>
          <tbody>
            {users.map((user) => (
              <tr key={user._id}>
                <td>{user._id}</td>
                <td>{user.name}</td>
                <td>{user.email}</td>
                <td>{user.isSuperAdmin ? "YES" : "NO"}</td>
                <td>{user.isAdmin ? "YES" : "NO"}</td>
                <td>{user.isStockKeeper ? "YES" : "NO"}</td>
                <td>
                  <button type="button" className="userlist-btn">
                    Edit
                  </button>
                  <button
                    type="button"
                    className="userlist-btn"
                    onClick={() => deleteHandler(user)}
                  >
                    Delete
                  </button>
                </td>
              </tr>
            ))}
          </tbody>
        </div>
      )}
    </div>
  );
}
