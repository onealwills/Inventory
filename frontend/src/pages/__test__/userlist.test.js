import React from "react";
import { render, fireEvent, act, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import UserListPage from "../UserListPage";

const mockStore = configureMockStore([thunk]);

describe("UserListPage", () => {
  let store;
  let wrapper;
  let mockDispatch;
  const mockUsers = [
    {
      _id: "123",
      name: "John Doe",
      email: "johndoe@example.com",
      isSuperAdmin: false,
      isAdmin: true,
      isStockKeeper: false,
    },
    // {
    //   _id: "456",
    //   name: "Jane Doe",
    //   email: "janedoe@example.com",
    //   isSuperAdmin: true,
    //   isAdmin: false,
    //   isStockKeeper: true,
    // },
  ];

  const mockState = {
    userList: { users: mockUsers, loading: false, error: null },
    userDelete: {
      loading: false,
      error: null,
      success: false,
    },
  };

  beforeEach(() => {
    mockDispatch = jest.fn();
    store = mockStore({ ...mockState });
    store.dispatch = mockDispatch;

    // mockDispatch.mockReturnValueOnce({
    //   payload: { userId: "123" },
    //   type: "USER_DELETE_REQUEST",
    // });

    wrapper = render(
      <Provider store={store}>
        <UserListPage />
      </Provider>
    );
  });

  it("should render the UserListPage component", () => {
    expect(wrapper.container).toMatchSnapshot();
  });

  it("should display the list of users", () => {
    const userList = wrapper.getAllByRole("row");
    expect(userList.length).toBe(2);
  });
  it("should dispatch the deleteUser action when the delete button is clicked", async () => {
    screen.debug();
    const deleteBtn = wrapper.getByText("Delete");
    fireEvent.click(deleteBtn);
    await act(async () => {
      window.confirm = jest.fn().mockImplementation(() => true);
    });
    console.log("mock dispatch>>>", mockDispatch);
    expect(mockDispatch).toHaveBeenCalled();
  });

  it("should not dispatch the deleteUser action when the user cancels the confirmation", async () => {
    const deleteBtn = wrapper.getByText("Delete");
    fireEvent.click(deleteBtn);
    await act(async () => {
      window.confirm = jest.fn().mockImplementation(() => false);
    });
    expect(mockDispatch).not.toHaveBeenCalledWith();
  });
});
