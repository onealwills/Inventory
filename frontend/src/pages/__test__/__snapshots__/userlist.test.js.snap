// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserListPage should render the UserListPage component 1`] = `
<div>
  <div>
    <h1>
      Users List
    </h1>
    <div>
      <thead>
        <tr>
          <th>
            ID
          </th>
          <th>
            NAME
          </th>
          <th>
            EMAIL
          </th>
          <th>
            IS SUPER ADMIN
          </th>
          <th>
            IS ADMIN
          </th>
          <th>
            IS STOCK KEEPER
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          role="row"
        >
          <td>
            123
          </td>
          <td>
            John Doe
          </td>
          <td>
            johndoe@example.com
          </td>
          <td>
            NO
          </td>
          <td>
            YES
          </td>
          <td>
            NO
          </td>
          <td>
            <button
              class="userlist-btn"
              type="button"
            >
              Edit
            </button>
            <button
              class="userlist-btn"
              type="button"
            >
              Delete
            </button>
          </td>
        </tr>
      </tbody>
    </div>
  </div>
</div>
`;
